scripts/db_migrator.py

```diff
@@ -53,15 +57,41 @@ def __init__(self, namespace, socket=None):
         self.TABLE_KEY       = 'DATABASE'
         self.TABLE_FIELD     = 'VERSION'
 
+        # load config data from golden_config_db.json
+        self.__golden_config_data = None
+        try:
+            if os.path.isfile(GOLDEN_CFG_FILE):
+                with open(GOLDEN_CFG_FILE) as f:
+                    golden_data = json.load(f)
+                    if namespace is None:
+                        self.__golden_config_data = golden_data
+                    else:
+                        if namespace == DEFAULT_NAMESPACE:
+                            config_namespace = "localhost"
+                        else:
+                            config_namespace = namespace
+                        self.__golden_config_data = golden_data[config_namespace]
+        except Exception as e:
+            log.log_error('Caught exception while trying to load golden config: ' + str(e))
+            pass
         # load config data from minigraph to get the default/hardcoded values from minigraph.py
-        # this is to avoid duplicating the hardcoded these values in db_migrator
-        self.minigraph_data = None
+        self.__minigraph_data = None
         try:
             if os.path.isfile(MINIGRAPH_FILE):
-                self.minigraph_data = parse_xml(MINIGRAPH_FILE)
+                self.__minigraph_data = parse_xml(MINIGRAPH_FILE)
         except Exception as e:
             log.log_error('Caught exception while trying to parse minigraph: ' + str(e))
             pass
+        # When both golden config and minigraph exists, override minigraph config with golden config
+        # config_src_data is the source of truth for config data
+        # this is to avoid duplicating the hardcoded these values in db_migrator
+        self.config_src_data = None
+        if self.__minigraph_data:
+            self.config_src_data = self.__minigraph_data
```

Comment:
>__minigraph_data [](http://example.com/codeflow?start=40&length=16)

This is shallow copy. So any future change on right side will also appear on left side.
I understand it is not efficient to make `minigraph_data` immutable. So instead, you can call it `config_src_data`, and override it.

`__xxx` convention does not help your case, it is just make the member super private.

Reply:
Fixed

Reply:
For reused function, add an argument to update_config, and update_config will use deepcopy by default, but db migrator will use shallow copy.
For db migrator, add a method to generate config_src_data. golden_config_data and minigraph_data are local variables, so it's impossible to alter them and then we can prevent unintended change.